### ISSUE TITLE
Dependency Management

### ISSUE THREAD

**Issue by benjamin-kirkbride** [Akuli/porcupine#1311]
Why do we have a requirements.txt, requirements-dev.txt, and a `dependencies` key in the `pyproject.toml`?

I would be happy to contribute a PR to move everything to `pyproject.toml` using `setuptools` as the build system.

**Comment by Akuli** [comment]
Porcupine used setuptools before, and I changed to flit in #1188. I don't remember why. Maybe I disliked having a `setup.py` file? Maybe flit seemed like a simpler "just works" approach to me? Maybe I didn't want to have anything to do with the complex mess of hacks that setuptools has become? Maybe it was a combination of these small reasons?

The purpose of `requirements.txt` and `requirements-dev.txt` is installing dependencies for development. I don't want to make contributors learn about flit or `pip install -e .` trickery, and most contributors are already familiar with `requirements.txt`. There is a test that fails if `requirements.txt` doesn't match the flit config, so having the same info in two places doesn't really cause problems.

That said, we could combine `requirements.txt` and `requirements-dev.txt` into a single file. When grepping for them I don't see anything that needs only one of two. Also, if setuptools works better for you for whatever purpose, we can switch back to it.

**Comment by benjamin-kirkbride** [comment]
> Maybe I disliked having a setup.py file?

You no longer need a setup.py to use `setuptools` as the build system, you can just have a single `pyproject.toml`

I don't know much about flit though, so maybe it's a non-issue.

> The purpose of requirements.txt and requirements-dev.txt is installing dependencies for development. I don't want to make contributors learn about flit or pip install -e .

I am not sure what the concern is here, my thought is simply document `pip install -e ".[dev]"` and everyone is good.

This is mostly prompted by https://github.com/Akuli/porcupine/issues/1312 - I would like to just install the package as editable and have pytest handle that well.

**Comment by Akuli** [comment]
Again, why do you want to install the package to develop it While it works, it just seems unnecessarily confusing. Everyone understands why you install dependencies, but installing the thing you work on is just weird (unless it is an operating system kernel or something).

**Comment by Akuli** [comment]
> You no longer need a setup.py to use setuptools as the build system, you can just have a single pyproject.toml

Apparently this is new in setuptools 61.0.0, so I imagine it would fail with a confusing error message if you have an older version of setuptools. For example, I have setuptools 52.0.0. So if we decide to do it this way, I would like to wait a few years first.

**Comment by benjamin-kirkbride** [comment]
Is telling people to update setuptools in their venv any different from asking them to install `flit` first?

**Comment by benjamin-kirkbride** [comment]
> why do you want to install the package to develop it

It minimizes the differences between how you develop and how users run the application. It removes the need for hacks like how the pytest.ini is set up.

See https://github.com/pythonarcade/arcade/blob/development/CONTRIBUTING.rst#installing-arcade-in-editable-mode and https://github.com/pythonarcade/arcade/blob/development/pyproject.toml for an example of how a project with a lot of contributors handles using nothing but setuptools and a pyproject.toml to handle development. This is a really common pattern now that pyproject.toml has matured and setuptools has full support.

All that said, I'm more advocating for `pip install ".[dev]"` being a supported way to be set up to do dev than I am looking to use `setuptools` specifically.

**Comment by Akuli** [comment]
**tl;dr:** We could try to get both ways to work simultaneously. I really like the simplicity and familiarity of `pip install -r requirements.txt` and don't really see any real advantages in doing it differently, but I also don't see why not support `".[dev]"` too.

You can ignore the rest of this, but I did try to explain my reasoning and point out a couple misunderstandings.

---

>Is telling people to update setuptools in their venv any different from asking them to install flit first?

I don't tell people to install `flit`. Pip does it automagically when you run `pip install https://github.com/Akuli/porcupine/archive/v2023.06.27.zip` or similar, because of `[build-system]` in `pyproject.toml`. On the other hand, I would need to tell people to upgrade `setuptools`. This isn't really a big deal, because I already tell people to upgrade `pip`.

>It minimizes the differences between how you develop and how users run the application.

I know that this is the theory, but of the hundreds of Porcupine bugs I have found, I'm pretty sure this would not have helped me with any of them. It is not solving a problem that I have. On the other hand, `PorcupineSetup_X.Y.Z.exe` installers work quite differently from the dev setup anyway (see `launcher/main.c`, `launcher/launch.pyw`, `scripts/build-exe-installer.py`), so there will always be big differences between how devs and users run Porcupine.

>It removes the need for hacks like how the pytest.ini is set up.

I think this is the only such hack. It is a hack though. Ideally I would tell pytest to just find all doctests in the `porcupine` module, but then it tries to import `porcupine/plugins/windows_run.py` which fails because that file is not meant to be imported. I am confused about how installing the project would help.

**Comment by benjamin-kirkbride** [comment]
I think your rationale is fair and your conclusion is a reasonable compromise.

> but then it tries to import porcupine/plugins/windows_run.py which fails because that file is not meant to be imported

Can you expand on this part? Is there a separate issue for that?

**Comment by Akuli** [comment]
Will expand in #1312.

**Comment by Akuli** [comment]
I think the only remaining part of this issue is supporting `pip install -e .` in Porcupine, in addition to the existing setup. You can add support for that if you want, and I will review and merge if it seems reasonable, but I don't really want to keep an issue open about it. I would probably feel different if multiple people requested the same thing.
